### PR TITLE
ENH: use pd.Grouper for tracking quality

### LINKED ISSUE #258: How to use temporal_tracking_quality to delete bad tracking days
Hey, 
I have a GPS tracking dataset and want to filter days that have a low tracking coverage. I noticed that the (pretty nice) function `ti.analysis.tracking_quality.temporal_tracking_quality` does exactly what I want ;-)
Except that I don't know how I can use the results for filtering. The problem is that the function does not return a date but a relative count of days since the start of tracking (I assume). Therefore I can not use the result to identify bad tracking days in my original dataset. (see the screenshot below). 
I think it would be better if the function would return a datetime index instead of a relativ day count. This is done automatically when using a pandas datetime grouper as in the modal split function [here](https://github.com/mie-lab/trackintel/blob/master/trackintel/analysis/modal_split.py#L69). Maybe using that would be an option.

image

### PR DESCRIPTION
**Pull request by hong2223**
closes #258

main changes:
- We use pd.Grouper to group days or weeks and return the datatime object. The granularity == "day" and "week" now only returns the absolute date. 
- As we have no relative days, we no longer fill in quality == 0 records (_get_all_quality() is now unused).
- Description added to the docstrings

### CODE REVIEW

**Comment by hong2223** [comment]
Please check if this PR achieve the functionalities we discussed @abcnishant007

File: trackintel/analysis/tracking_quality.py [hunk 1]
```diff
@@ -40,17 +40,33 @@ def temporal_tracking_quality(source, granularity="all", max_iter=60):
...
+
+    - ``weekday``
+
+        - time extent: the whole day (86400 sec) * number of tracked weeks for each user for each user
+        - columns: ``['user_id', 'weekday', 'quality']``
+
+    - ``hour``:
+
+        - time extent: the whole hour (3600 sec) * number of tracked days for each user
+        - columns: ``['user_id', 'hour', 'quality']``
```
Comment (abcnishant007) [inline trackintel/analysis/tracking_quality.py:70]: +1 to the docstring update, this summarises everything.
Also, since you did not change the order of the `quality` column, other test functions need not be updated as well (for example, I had to make it -2 at some places instead of -1). Now it can be -1 throughout. 

I think this is a better approach than what I was doing (returning both items: relative and absolute.)

File: tests/analysis/test_tracking_quality.py [hunk 2]
```diff
@@ -69,8 +69,8 @@ def test_tracking_quality_day(self, testdata_sp_tpls_geolife_long):
         # test if the result of the user agrees
         quality = ti.analysis.tracking_quality.temporal_tracking_quality(sp_tpls, granularity="day")
 
-        assert quality_manual == quality.loc[(quality["user_id"] == 0) & (quality["day"] == 0), "quality"].values[0]
-        assert (quality["quality"] < 1).all()
+        assert quality_manual == quality.iloc[0]["quality"]
+        assert (quality["quality"] <= 1).all()
```
Comment (abcnishant007) [inline tests/analysis/test_tracking_quality.py:73]: Why did we need `<=` instead of `<` that was there before?

File: tests/analysis/test_tracking_quality.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ def test_tracking_quality_week(self, testdata_sp_tpls_geolife_long):
         quality = ti.analysis.tracking_quality.temporal_tracking_quality(sp_tpls, granularity="week")
 
         assert quality_manual == quality.loc[(quality["user_id"] == 0), "quality"].values[0]
-        assert (quality["quality"] < 1).all()
+        assert (quality["quality"] <= 1).all()
```
Comment (abcnishant007) [inline tests/analysis/test_tracking_quality.py:95]: same as previous comment

File: tests/analysis/test_tracking_quality.py [hunk 4]
```diff
@@ -221,6 +221,48 @@ def test_non_positive_duration_warning(self):
...
+
+        quality = ti.analysis.tracking_quality.temporal_tracking_quality(sp, granularity="day")
+        # get the "date" of the last record and compare to the last "date" in data
+        assert quality.values[-1][-2] == (t1 + ten_days - pd.Timedelta(days=1))
+
+    def test_last_day_same_as_last_week_in_record(self):
+        """Test the take the last tracking record's day and assert it is the same as the
+        last absolute date in the quality."""
+        p1 = Point(8.5067847, 47.4)
+        t1 = pd.Timestamp("1971-01-01 00:00:00", tz="utc")
```
Comment (abcnishant007) [inline tests/analysis/test_tracking_quality.py:249]: Should we add another test case where we have a time stamp starting NOT at `00:00:00`so that we can be sure that it works as expected for non-boundary cases? 
Reminder: I am referring to the issue which I had in my solution, extra day due to starting at 00:00:00

**Review by abcnishant007 — COMMENTED** [review]
Added some pointers to draw your attention to some minor points. Otherwise this is good!

File: tests/analysis/test_tracking_quality.py (hunk 2 above)
Comment (hong2223) [reply tests/analysis/test_tracking_quality.py:89]: This relates to a bug we introduced before, check #221. So quality theoretically could take the value 1. The change here is just for the "correctness" and did not influence test results

**Review by hong2223 — COMMENTED** [review]
Thanks for taking a look. I separated the function for generating test records and added one test case for a time starting in the middle of the day. Shall be ready for review

**Comment by abcnishant007** [comment]
Thanks for making the changes. 

Looking at the report, I don't see coverage being reduced. Is the `codecov` red flag due to some codecov bug? If yes, please feel free to merge.

**Comment by hong2223** [comment]
I am not sure what causes the failure of our tests. We need to fix this before merging

**Comment by henrymartin1** [comment]
> 
> 
> I am not sure what causes the failure of our tests. We need to fix this before merging

I restarted the test job. There was a similar problem on the master branch that was solved by re-running the tests ;-)

**Comment by abcnishant007** [comment]
> > I am not sure what causes the failure of our tests. We need to fix this before merging
> 
> I restarted the test job. There was a similar problem on the master branch that was solved by re-running the tests ;-)

That used to work but didn't work this time it seems. I will check if pushing another dummy commit fixes it.
